src/viam/sdk/robot/client.cpp

```diff
@@ -57,6 +53,146 @@ using viam::robot::v1::Status;
 // NOLINTNEXTLINE
 const std::string kStreamRemoved("Stream removed");
 
+// TODO: add a traits class for proto to type and back conversion
+DiscoveryQuery RobotClient::discovery_query::to_proto() const {
+    DiscoveryQuery proto;
+    *proto.mutable_subtype() = subtype;
+    *proto.mutable_model() = std::move(model);
+    return proto;
+}
+
+RobotClient::discovery_query RobotClient::discovery_query::from_proto(const DiscoveryQuery& proto) {
+    discovery_query query;
+    query.subtype = proto.subtype();
+    query.model = proto.model();
+    return query;
+}
+
+bool operator==(const RobotClient::discovery_query& lhs, const RobotClient::discovery_query& rhs) {
+    return lhs.subtype == rhs.subtype && lhs.model == rhs.model;
+}
+
+Discovery RobotClient::discovery::to_proto() const {
+    Discovery proto;
+    *proto.mutable_query() = query.to_proto();
+    *proto.mutable_results() = map_to_struct(results);
+    return proto;
+}
+
+RobotClient::discovery RobotClient::discovery::from_proto(const Discovery& proto) {
+    discovery discovery;
+    discovery.query = discovery_query::from_proto(proto.query());
+    discovery.results = struct_to_map(proto.results());
+    return discovery;
+}
+
+bool operator==(const RobotClient::discovery& lhs, const RobotClient::discovery& rhs) {
+    return lhs.query == rhs.query && map_to_struct(lhs.results).SerializeAsString() ==
```

Comment:
Why do `results` get turned back into `proto Struct`'s in order to compare them with `SerializeAsString`? I'd expect `(*lhs.results == *rhs.results)` to work, since then you are comparing unordered maps.

Ohhh.... I think I see. It won't work because it will end up comparing the embedded `shared_ptr<ProtoType>` rather than the pointed-to `ProtoType`. That's unfortunate.

I think we should prioritize fixing things up around `AttributeMap` / `ProtoType`. I think if we do that we will end up back here because I'd expect to see `map_to_struct` and `struct_to_map` change.

OK to go ahead with what you have here unless @stuqdog disagrees.

Reply:
No disagreement here. Agreed also that `AttributeMap` could use another look. Do you recall offhand if we made a ticket for that already?

Reply:
No, I don't think we have. Could you make one?

Reply:
Done. https://viam.atlassian.net/browse/RSDK-6636